### PR TITLE
Bucket2.0

### PR DESCRIPTION
**Pull request by kmdeck**
This PR adds in a test case showing a coupled atmos, land, ocean, and sea ice model. The bucket model in place of the slab land model, which is how it differs from https://github.com/CliMA/ClimaCoupler.jl/pull/50.

We've also set it up so that the specific humidity at the surface uses the correct land value, and set the RRTMGP fields for surface temp and albedo correctly based on the surface values.

A future PR will update the version of ClimaAtmos we are using so that we can read the precipitation from the atmos cache, and pass it to the land model, and so that we are using topography in the atmos model.

A future PR will also implement a coupler interface as this interface is currently hard coded to some extent to the models being used here.

co-authors:
Ben Mackay @jb-mackay
Lenka Novak @LenkaNovak
Zhao-yi Shen @szy21
Charlie @charleskawczynski

### CODE REVIEW

**Comment by kmdeck** [comment]
As of "conserves energy" push, commit # [57c8b49](https://github.com/CliMA/ClimaCoupler.jl/pull/57/commits/57c8b49910289396b8727597603bbbf7c915067b)
total_energy_log
total_energy

**Comment by kmdeck** [comment]
As of [added bucket capability minus albedo](https://github.com/CliMA/ClimaCoupler.jl/pull/57/commits/95f126e461d51c19fb550c5a1b7a7e9ce485fa62)

Bucket model conserves energy
total_energy_log_bucket_x
total_energy_bucket_x

**Comment by kmdeck** [comment]
As of [added q_sfc and artifacts](https://github.com/CliMA/ClimaCoupler.jl/pull/57/commits/0d82e109fe9cd9c42764d34541b109b3df478f7b)
total_energy_log_bucket_x
total_energy_bucket_x

**Comment by kmdeck** [comment]
30 day run:
total_energy_log_bucket_x
total_energy_bucket_x

Videos:
\rho q t/\rho at first level:
https://user-images.githubusercontent.com/65979205/173682567-38f1c211-2996-4d00-897e-6b9f437cce0d.mp4
\rho e at first level
https://user-images.githubusercontent.com/65979205/173682574-75e3e26b-ff17-48b1-9b8b-7668898ddae1.mp4
wind u at first level
https://user-images.githubusercontent.com/65979205/173682581-04868dff-b1b7-4e07-aa94-e67cf9c70cfb.mp4
Earth T 
https://user-images.githubusercontent.com/65979205/173682587-b4a85988-79a4-49b3-976d-4595f4eaa2af.mp4
Bucket Moisture - 0 plotted for ocean and ice.
https://user-images.githubusercontent.com/65979205/173682593-e9f219f6-4bd5-45bd-8c56-998f9ac0123a.mp4

**Comment by LenkaNovak** [comment]
Looks good! In the next PR with the updated `ClimaAtmos` interface (with precipitation + topography), we should
    - add conductive sea-ice flux
    - improve the docs
    - add a better interface for the mask (similar to what was in Bucket1.0)
    - double check the `end - half` issue in the conservation check that Zhaoyi mentioned (btw, `end + half` gives an error)
    - ensure we remove (only) the unnecessary tmp files

File: experiments/ClimaCore/bc-wave-slab/slab.jl [hunk 1]
```diff
@@ -58,7 +58,7 @@ end
 
 function slab_init(
```
Comment (jb-mackay) [inline experiments/ClimaCore/bc-wave-slab/slab.jl:59]: Files modified in `experiments/ClimaCore/...` are unrelated to this PR and should be removed/reverted.

File: experiments/AMIP/moist_mpi_earth/bucket/bucket_init.jl [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+struct BucketSimulation{P, Y, D, I}
+    params::P
+    Y_init::Y
+    domain::D
+    integrator::I
+end
+
+# init simulation
+function initialize(model::BucketModel, space::ClimaCore.Spaces.AbstractSpace)
```
Comment (jb-mackay) [inline experiments/AMIP/moist_mpi_earth/bucket/bucket_init.jl:24]: Does it make sense to add this method to ClimaLSM @kmdeck? Probably something we will use again. We can merge this PR like this and open a PR for LSM, order doesn't really matter.

File: experiments/AMIP/moist_mpi_earth/coupler_utils/conservation_checker.jl [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+function check_conservation(
+    cs::OnlineConservationCheck,
+    coupler_sim,
+    atmos_sim = nothing,
+    land_sim = nothing,
+    ocean_sim = nothing,
+    seaice_sim = nothing,
+    radiation = true,
+)
+    z = parent(Fields.coordinate_field(face_space).z)
```
Comment (jb-mackay) [inline experiments/AMIP/moist_mpi_earth/coupler_utils/conservation_checker.jl:22]: `face_space` is not defined here - it is probably passed as a global currently.

File: experiments/AMIP/moist_mpi_earth/coupler_utils/conservation_checker.jl [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    ice_mask(u_ice_1, univ_mask) = (univ_mask ≈ FT(-2) ? u_ice_1 : FT(0))
+    ocean_mask(u_ocn_1, univ_mask) = (univ_mask ≈ FT(0) ? u_ocn_1 : FT(0))
+
+    # Save land
+    parent(u_lnd) .= land_mask.(parent(u_lnd), univ_mask)
+    land_e = land_sim !== nothing ? sum(get_land_energy(land_sim, u_lnd)) ./ Δz_bot : FT(0)
+    land_sim !== nothing ? push!(cs.ρe_tot_land, land_e) : nothing
+
+    # Save ocean and ice
+    if (prescribed_sst !== true) && (prescribed_sic == true)
```
Comment (jb-mackay) [inline experiments/AMIP/moist_mpi_earth/coupler_utils/conservation_checker.jl:78]:
```suggestion
    if !prescribed_sst && prescribed_sic
```

File: experiments/AMIP/moist_mpi_earth/coupler_utils/conservation_checker.jl [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+        ocean_e = ocean_sim !== nothing ? sum(get_slab_energy(ocean_sim, u_ocn)) ./ Δz_bot : FT(0)
+        seaice_e = seaice_sim !== nothing ? sum(get_slab_energy(seaice_sim, u_ice)) ./ Δz_bot : FT(0)
+
+        # save to coupler cache
+        ocean_sim !== nothing ? push!(cs.ρe_tot_ocean, ocean_e) : nothing
+        seaice_sim !== nothing ? push!(cs.ρe_tot_seaice, seaice_e) : nothing
+    end
+
+    if (prescribed_sic == false) # rm? dynamic sea ice + associated ocean are WIP (different PR)
```
Comment (jb-mackay) [inline experiments/AMIP/moist_mpi_earth/coupler_utils/conservation_checker.jl:90]:
```suggestion
    if !prescribed_sic # rm? dynamic sea ice + associated ocean are WIP (different PR)
```

File: experiments/AMIP/moist_mpi_earth/bucket/bucket_init.jl (hunk 2 above)
Comment (kmdeck) [reply experiments/AMIP/moist_mpi_earth/bucket/bucket_init.jl:24]: agreed! I'd say let's wait on moving this to LSM until we have another land model we want to couple & the exact form of the Simulation struct is finalized. perhaps we will have a LandSimulation with an addition field = model?

File: experiments/AMIP/moist_mpi_earth/coupler_utils/conservation_checker.jl (hunk 3 above)
Comment (kmdeck) [reply experiments/AMIP/moist_mpi_earth/coupler_utils/conservation_checker.jl:22]: yes...we'll have to revisit a few of those! FT is also a global, and mask.

File: experiments/AMIP/moist_mpi_earth/slab_ice/slab_init.jl [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+# sea-ice parameters
+struct IceSlabParameters
+    h::FT # sea ice height
+    ρ::FT
+    c::FT
+    T_base::FT
+    z0m::FT
+    z0b::FT
+    T_freeze::FT # temperature at freezing point [K] 
+    k_ice::FT   # thermal conductivity of ice [W / m / K]
```
Comment (kmdeck) [inline experiments/AMIP/moist_mpi_earth/slab_ice/slab_init.jl:12]: Note: the conservation tests all assume k_ice and k_land are zero, so that there is no conductive heat flux at the bottom.

File: experiments/AMIP/moist_mpi_earth/coupler_utils/conservation_checker.jl (hunk 5 above)
Comment (kmdeck) [reply experiments/AMIP/moist_mpi_earth/coupler_utils/conservation_checker.jl]: removing!

File: experiments/AMIP/moist_mpi_earth/coupler_utils/conservation_checker.jl (hunk 4 above)
Comment (kmdeck) [reply experiments/AMIP/moist_mpi_earth/coupler_utils/conservation_checker.jl]: good point!

**Comment by kmdeck** [comment]
Make sure prescribed SST still runs, and we should be able to generate videos for it

**Comment by kmdeck** [comment]
As of [cleaned up push pull a bit](https://github.com/CliMA/ClimaCoupler.jl/pull/57/commits/97bf45e03547a18e0f990041bd22f0962c8c2cf5)
still conserving energy for slab and bucket versions of land, eg:

total_energy_log_bucket

**Comment by kmdeck** [comment]
> Looks good! In the next PR with the updated `ClimaAtmos` interface (with precipitation + topography), we should - add conductive sea-ice flux - improve the docs - add a better interface for the mask (similar to what was in Bucket1.0) - double check the `end - half` issue in the conservation check that Zhaoyi mentioned (btw, `end + half` gives an error) - ensure we remove (only) the unnecessary tmp files

I think the nz -half is fine - the calculation gives nz = 11 (for 10 layers), so 11-half = 10.5 = what we want for the upper face?

**Comment by szy21** [comment]
> 
> I think the nz -half is fine - the calculation gives nz = 11 (for 10 layers), so 11-half = 10.5 = what we want for the upper face?

Oh I didn’t know nz is 11. I think that’s fine then.

**Comment by kmdeck** [comment]
> > I think the nz -half is fine - the calculation gives nz = 11 (for 10 layers), so 11-half = 10.5 = what we want for the upper face?
> 
> Oh I didn’t know nz is 11. I think that’s fine then.

we can rename it!

File: experiments/AMIP/moist_mpi_earth/coupler_utils/general_helper.jl [hunk 7]
```diff
@@ -0,0 +1,20 @@
+# most of these functions are temporary helpers until upstream issues are resolved
+
+# TODO: unify with coupler interface
+struct CouplerSimulation{I, F, B, T, M}
+    Δt::I
+    t::F
+    boundary_space::B
+    FT::T
+    mask::M
+end
```
Comment (jb-mackay) [inline experiments/AMIP/moist_mpi_earth/coupler_utils/general_helper.jl:10]:
```suggestion
struct CouplerSimulation{FT, I, F, B, M}
    Δt::I
    t::F
    boundary_space::B
    mask::M
end
```

we can then access `FT` by doing things like `foo(couplersim::CouplerSimulation{FT},  ...) where {FT}`

File: experiments/AMIP/moist_mpi_earth/coupler_utils/general_helper.jl [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+get_u(sim, t) = Geometry.UVVector.(sim.integrator.sol.u[t].c.uₕ).components.data.:1
```
Comment (jb-mackay) [inline experiments/AMIP/moist_mpi_earth/coupler_utils/general_helper.jl:12]: I'd like this to have a more specific name since `u` could refer to `integrator.sol.u`

File: experiments/AMIP/moist_mpi_earth/coupler_utils/masker.jl [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    mask = ncreader_rll_to_cgll(FT, infile, varname, ne = ne, R = R, Nq = Nq, outfile = outfile)
+    mask = clean_mask.(FT, mask)
+    # LandSeaMask(infile, ne, Nq, R, mask)
+end
+
+"""
+clean_mask(FT, mask)
+- convert to integer values after interpolation (but keep type as floats foe easier calculation (TODO))
+"""
+clean_mask(FT, mask) = mask > FT(0.7) ? FT(1) : FT(0)
```
Comment (jb-mackay) [inline experiments/AMIP/moist_mpi_earth/coupler_utils/masker.jl:25]: How did we determine the 0.7 threshold?

File: experiments/AMIP/moist_mpi_earth/coupler_utils/masker.jl [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+combine_surface(mask, sfc_1, sfc_2, sfc_3, value1 = -0.5, value2 = 0.5) =
+    (mask < FT(value1) ? sfc_3 : FT(0)) +
+    ((mask >= FT(value1) && (mask <= FT(value2))) ? sfc_2 : FT(0)) +
+    (mask > FT(value2) ? sfc_1 : FT(0))
+
+"""
+apply_mask(mask, condition, yes, no, value = 0.5) 
+- apply mask mased on a threshold value in the mask
+"""
+apply_mask(mask, condition, yes, no, value) = condition(mask, value) ? yes : no
```
Comment (jb-mackay) [inline experiments/AMIP/moist_mpi_earth/coupler_utils/masker.jl:42]: `yes` and `no` are not great argument names, it's unclear what they should be

File: experiments/AMIP/moist_mpi_earth/slab/slab_utils.jl [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+function get_land_roughness(slab_sim::SlabSimulation)
+    return slab_sim.integrator.p.params.z0m, slab_sim.integrator.p.params.z0b
+end
+
+land_albedo(slab_sim::SlabSimulation) = slab_sim.params.α
+
+function get_land_q(slab_sim::SlabSimulation, atmos_sim, T_land, ρ_sfc)
+    return TD.q_vap_saturation_generic.(atmos_sim.integrator.p.params, T_land, ρ_sfc, TD.Liquid())
+end
```
Comment (jb-mackay) [inline experiments/AMIP/moist_mpi_earth/slab/slab_utils.jl:26]: We can remove these as they are defined for the bucket now in `bucket_init`

Comment (jb-mackay) [inline experiments/AMIP/moist_mpi_earth/slab/slab_utils.jl:26]:
```suggestion

```
We can remove these as they are defined for the bucket now in bucket_init

Comment (kmdeck) [reply experiments/AMIP/moist_mpi_earth/slab/slab_utils.jl:26]: do we want to preserve being able to use a slab land instead of the bucket? it's a minor difference - the bucket energy part is a slab, but it is based on a different repo, etc. These methods would be necessary if we want to retain both.

File: experiments/AMIP/moist_mpi_earth/vtk_helper.jl [hunk 12]
```diff
@@ -0,0 +1,58 @@
+## Plot VTK output on the 2D cubed-sphere, building on from ClimaCoreWorkshops
+using ClimaCoreVTK
```
[the remark this answers is not in the document]
Comment (simonbyrne) [reply experiments/AMIP/moist_mpi_earth/vtk_helper.jl:2]: anything specifically?

File: experiments/AMIP/moist_mpi_earth/coupler_utils/general_helper.jl (hunk 8 above)
Comment (kmdeck) [reply experiments/AMIP/moist_mpi_earth/coupler_utils/general_helper.jl]: not being used :) removed it

File: experiments/AMIP/moist_mpi_earth/coupler_utils/masker.jl (hunk 9 above)
Comment (kmdeck) [reply experiments/AMIP/moist_mpi_earth/coupler_utils/masker.jl]: I made it a parameter, not sure how it is chosen

File: experiments/AMIP/moist_mpi_earth/coupler_utils/masker.jl (hunk 10 above)
Comment (kmdeck) [reply experiments/AMIP/moist_mpi_earth/coupler_utils/masker.jl]: Edited this - removed the "no" argument, since the point of this is to zero out a field where the mask is zero. so now it is apply_mask(mask, condition, field, value).

File: experiments/AMIP/moist_mpi_earth/bucket/bucket_init.jl (hunk 2 above)
Comment (kmdeck) [reply experiments/AMIP/moist_mpi_earth/bucket/bucket_init.jl:24]: i misread this! the #init simulation comment threw me off. yeah let's move this into ClimaLSM

File: experiments/AMIP/moist_mpi_earth/coupler_driver.jl [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+    slab_sim = bucket_init(FT;tspan =  FT.(tspan), dt = FT(Δt_cpl), space = boundary_space, saveat = FT(saveat))
+end
+
+prescribed_sst = false
+if prescribed_sst
+    println("No ocean sim - do not expect energy conservation")
+    SST =
+        ncreader_rll_to_cgll_from_space(FT, time_slice_ncfile(sst_data), "SST", boundary_space, outfile = "sst_cgll.nc")  # a sample SST field
+
+    SST = swap_space!(SST, axes(mask)) .* (abs.(mask .- 1)) .+ FT(273.15)
```
Comment (kmdeck) [inline experiments/AMIP/moist_mpi_earth/coupler_driver.jl:73]: no need to zero out the SST and SIC fields here, since we account for the masks properly when computing fluxes

File: experiments/AMIP/moist_mpi_earth/coupler_driver.jl [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+
+slab_ice_sim =
+    slab_ice_init(FT; tspan = tspan, dt = Δt_cpl, space = boundary_space, saveat = saveat, ice_mask = ice_mask)
+
+
+# init coupler
+coupler_sim = CouplerSimulation(FT(Δt_cpl), integrator.t, boundary_space, FT, mask)
+include("./push_pull.jl")
+# init conservation info collector
+CS = OnlineConservationCheck([], [], [], [], [], [])
```
Comment (kmdeck) [inline experiments/AMIP/moist_mpi_earth/coupler_driver.jl:103]: maybe create a flag for energy checking..

File: experiments/AMIP/moist_mpi_earth/coupler_utils/general_helper.jl (hunk 7 above)
Comment (kmdeck) [reply experiments/AMIP/moist_mpi_earth/coupler_utils/general_helper.jl:10]: currently am just grabbing FT from eltype(Y), since couplersim isnt used by most methods. but good call for the future!

File: experiments/AMIP/moist_mpi_earth/mpi/mpi_postprocess.jl [hunk 15]
```diff
@@ -0,0 +1,37 @@
+
+# function mpi_collapse_solu()# (comms_ctx, horizontal_mesh, quad, z_max, z_elem, z_stretch, Y) # TODO: it's using the global scope here
+if is_distributed # replace sol.u on the root processor with the global sol.u
+
+    if ClimaComms.iamroot(comms_ctx)
+        global_h_space = make_horizontal_space(horizontal_mesh, quad, nothing)
```
Comment (kmdeck) [inline experiments/AMIP/moist_mpi_earth/mpi/mpi_postprocess.jl:6]: define FT here

File: experiments/AMIP/moist_mpi_earth/mpi/mpi_postprocess.jl [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+        global_Y_slab = DataLayouts.gather(comms_ctx, Fields.field_values(sol_slab.u[i].T_sfc))
+        if ClimaComms.iamroot(comms_ctx)
+            global_sol_u_atmos[i] = Fields.FieldVector(
+                c = Fields.Field(global_Y_c, global_center_space),
+                f = Fields.Field(global_Y_f, global_face_space),
+            )
+            global_sol_u_slab[i] = Fields.Field(global_Y_slab, global_h_space)
+        end
+    end
+    if ClimaComms.iamroot(comms_ctx)
```
Comment (kmdeck) [inline experiments/AMIP/moist_mpi_earth/mpi/mpi_postprocess.jl:28]: Im not sure this code supports having an ocean model, and ice model, or slab vs bucket land model.

@LenkaNovak can I make an issue to address this rather than do it in this PR? I havent worked with MPI at all.

File: experiments/AMIP/moist_mpi_earth/vtk_helper.jl (hunk 12 above)
[the remark this answers is not in the document]
Comment (kmdeck) [reply experiments/AMIP/moist_mpi_earth/vtk_helper.jl:2]: as far as I can tell this is not being used in this PR, Ill remove it @LenkaNovak

[the remark this answers is not in the document]
Comment (jb-mackay) [reply experiments/AMIP/moist_mpi_earth/vtk_helper.jl:2]: @simonbyrne referring to these `retype_space` functions converting spaces from F32 to F64 to work with `ClimaCoreVTK.writepvd`.

File: experiments/AMIP/moist_mpi_earth/mpi/mpi_postprocess.jl (hunk 15 above)
Comment (kmdeck) [reply experiments/AMIP/moist_mpi_earth/mpi/mpi_postprocess.jl:6]: this is also in global, leaving alone?

File: experiments/AMIP/moist_mpi_earth/mpi/mpi_postprocess.jl (hunk 16 above)
Comment (LenkaNovak) [reply experiments/AMIP/moist_mpi_earth/mpi/mpi_postprocess.jl:28]: This should work for several models (though it's not very efficient since all surface model components are masked global shells). Since we're only using these really simple ODE models for AMIP though, the slowdown shouldn't be too bad.

File: experiments/AMIP/moist_mpi_earth/Project.toml [hunk 17]
```diff
@@ -42,7 +42,6 @@ UnPack = "3a884ed6-31ef-47d7-9d2a-63182c4928ed"
 
 [compat]
 ClimaCore = "0.10.3"
-RRTMGP = "0.3.2"
```
Comment (charleskawczynski) [inline experiments/AMIP/moist_mpi_earth/Project.toml]: We actually need this unless we upgrade to the new clima parameter handling, but I think that may be better as a separate PR

**Comment by charleskawczynski** [comment]
I recommend reseting to `f85f27fa868789f096558760083059bf69f67c6c` and merging, the remaining changes could be applied in a separate PR. Upgrading to the latest versions of packages requires changing how we handle clima parameters, which is not trivial. These packages in include

 - RRTMGP v0.4.0
 - TurbulecenceConvection v0.33
 - Thermodynamics 0.9
 - CloudMicrophysics v0.6

and more. The list is actually pretty long, and the required changes are complicated, as you can see in the [PR](https://github.com/CliMA/ClimaAtmos.jl/pull/548) where we upgraded to latest packages in ClimaAtmos

**Comment by kmdeck** [comment]
Results after hooking up with latest clima atmos, precip fixed, coupler setting atmos BC:

qt lowest level:
https://user-images.githubusercontent.com/65979205/175792073-9c1c3d85-5fda-412a-ab77-85c45fe92da4.mp4

u lowest level:
https://user-images.githubusercontent.com/65979205/175792077-16f40323-da1a-4ee2-b633-23e0ddc5e0bc.mp4

earth T:
https://user-images.githubusercontent.com/65979205/175792084-b2879081-c95d-4c2b-bad4-ec917a0ca07b.mp4

bucket W:

https://user-images.githubusercontent.com/65979205/175792087-1e858d33-bf4d-43a3-93b5-2623da36f1e6.mp4

Energy error, fractional, of entire system:
total_energy_log_bucket

**Comment by kmdeck** [comment]
@szy21 @charleskawczynski @LenkaNovak @jb-mackay 

IMO we can merge this now. We're using the hot fix commit from ClimaAtmos, so the coupler is now setting the BC for clima atmos, and precipitation is nonzero by the end of the simulation, and land bucket water content doesnt seem unreasonable. To check the accuracy of this further, we probably need to have diagnostics set up. Land needs to make a callback function, we can implement/use the callback clima atmos already makes in the driver file.

Perhaps one small fix before merging is to figure out where the plots in buildkite get put/how to access those.

**Comment by LenkaNovak** [comment]
Agreed, thanks, Kat! After the issues on reverting and buildkite are sorted, let's merge and sort out diagnostics as part of the next interface-focused iteration. WRT buildkite plots, are you thinking of setting up something like what we had a while back for the GCM results where buildkite would send a selection of plots on Slack?

**Comment by charleskawczynski** [comment]
> EDIT: I just saw @charleskawczynski made a comment about reverting to a previous commit, since different packages are in different stages of working with clima params. Charlie, the code runs right now, though....using the changes you made to climaLSM to use the new parameters, and I believe using a commit of clima atmos that is using the new params as well e88fbea293bf6fb48a927f0459556d1de4dca0e9. so why revert?

Because we haven't gone through and actually applied the clima parameter changes. If you want more information I can look through and point out all of the things that are inconsistent. Basically, anything that is using CLIMAParameters is inconsistent with how their being used downstream.
